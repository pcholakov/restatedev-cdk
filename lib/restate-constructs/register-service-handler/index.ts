/*
 * Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */

import { Handler } from "aws-lambda/handler";
import { CloudFormationCustomResourceEvent } from "aws-lambda/trigger/cloudformation-custom-resource";
import fetch from "node-fetch";
import * as cdk from "aws-cdk-lib";
import { GetSecretValueCommand, SecretsManagerClient } from "@aws-sdk/client-secrets-manager";
import * as https from "https";
import { randomInt } from "crypto";

export interface RegistrationProperties {
  servicePath?: string;
  adminUrl?: string;
  serviceLambdaArn?: string;
  invokeRoleArn?: string;
  removalPolicy?: cdk.RemovalPolicy;
  authTokenSecretArn?: string;
}

type RegisterDeploymentResponse = {
  id?: string;
  services?: { name?: string; revision?: number }[];
};

const MAX_HEALTH_CHECK_ATTEMPTS = 5; // This is intentionally quite long to allow some time for first-run EC2 and Docker boot up
const MAX_REGISTRATION_ATTEMPTS = 3;

const INSECURE = true;

const DEPLOYMENTS_PATH = "deployments";
const DEPLOYMENTS_PATH_LEGACY = "endpoints"; // temporarily fall back for legacy clusters

/**
 * Custom Resource event handler for Restate service registration. This handler backs the custom resources created by
 * {@link LambdaServiceRegistry} to facilitate Lambda service handler discovery.
 */
export const handler: Handler<CloudFormationCustomResourceEvent, void> = async function(event) {
  console.log({ event });

  if (event.RequestType === "Delete") {
    // Since we retain older Lambda handler versions on update, we also leave the registered service alone. There may
    // be unfinished invocations that require it; in the future we would want to inform Restate that we want to
    // de-register the service, and wait for Restate to let us know that it is safe to delete the deployed Function
    // version from Lambda.

    // const props = event.ResourceProperties as RegistrationProperties;
    // if (props.removalPolicy === cdk.RemovalPolicy.DESTROY) {
    //   console.log(`De-registering service ${props.serviceLambdaArn}`);
    //   const controller = new AbortController();
    //   const id = btoa(props.serviceLambdaArn!); // TODO: we should be treating service ids as opaque
    //   const deleteCallTimeout = setTimeout(() => controller.abort("timeout"), 5_000);
    //   const deleteResponse = await fetch(`${props.adminUrl}/${DEPLOYMENTS_PATH}/${id}?force=true`, {
    //     signal: controller.signal,
    //     method: "DELETE",
    //     agent: INSECURE ? new https.Agent({ rejectUnauthorized: false }) : undefined,
    //   }).finally(() => clearTimeout(deleteCallTimeout));
    //
    //   console.log(`Got delete response back: ${deleteResponse.status}`);
    //   if (deleteResponse.status != 202) {
    //     throw new Error(`Removing service deployment failed: ${deleteResponse.statusText} (${deleteResponse.status})`);
    //   }
    // }

    console.warn("De-registering services is not supported currently. Previous version will remain registered.");
    return;
  }

  const props = event.ResourceProperties as RegistrationProperties;
  const authHeader = await createAuthHeader(props);

  let attempt;
  const controller = new AbortController();

  const healthCheckUrl = `${props.adminUrl}/health`;

  console.log(`Performing health check against: ${healthCheckUrl}`);
  attempt = 1;
  while (true) {
    const healthCheckTimeout = setTimeout(() => controller.abort("timeout"), 5_000);
    let healthResponse = undefined;
    let errorMessage = undefined;
    try {
      healthResponse = await fetch(healthCheckUrl, {
        signal: controller.signal,
        headers: authHeader,
        agent: INSECURE ? new https.Agent({ rejectUnauthorized: false }) : undefined,
      }).finally(() => clearTimeout(healthCheckTimeout));

      console.log(`Got health check response back: ${healthResponse.status}`);
      if (healthResponse.status >= 200 && healthResponse.status < 300) {
        break;
      }
      console.error(
        `Restate health check failed: ${healthResponse.statusText} (${healthResponse.status}; attempt ${attempt})`,
      );
    } catch (e) {
      errorMessage = (e as Error)?.message;
      console.error(`Restate health check failed: "${errorMessage}" (attempt ${attempt})`);
    }

    if (attempt >= MAX_HEALTH_CHECK_ATTEMPTS) {
      console.error(`Admin service health check failing after ${attempt} attempts.`);
      throw new Error(errorMessage ?? `${healthResponse?.statusText} (${healthResponse?.status})`);
    }
    attempt += 1;

    const waitTimeMillis = randomInt(2_000) + 2 ** attempt * 1_000; // 3s -> 6s -> 10s -> 18s -> 34s
    console.log(`Retrying after ${waitTimeMillis} ms...`);
    await sleep(waitTimeMillis);
  }

  let deploymentsUrl = `${props.adminUrl}/${DEPLOYMENTS_PATH}`;
  const registrationRequest = JSON.stringify({
    arn: props.serviceLambdaArn,
    assume_role_arn: props.invokeRoleArn,
  });

  let failureReason;
  console.log(`Triggering registration at ${deploymentsUrl}: ${registrationRequest}`);
  attempt = 1;
  while (true) {
    try {
      const registerCallTimeout = setTimeout(() => controller.abort("timeout"), 10_000);
      const registerDeploymentResponse = await fetch(deploymentsUrl, {
        signal: controller.signal,
        method: "POST",
        body: registrationRequest,
        headers: {
          "Content-Type": "application/json",
          ...authHeader,
        },
        agent: INSECURE ? new https.Agent({ rejectUnauthorized: false }) : undefined,
      }).finally(() => clearTimeout(registerCallTimeout));

      console.log(`Got registration response back: ${registerDeploymentResponse.status}`);

      if (registerDeploymentResponse.status == 404 && attempt == 1) {
        deploymentsUrl = `${props.adminUrl}/${DEPLOYMENTS_PATH_LEGACY}`;
        console.log(`Got 404, falling back to <0.7.0 legacy endpoint registration at: ${deploymentsUrl}`);
      }

      if (registerDeploymentResponse.status >= 200 && registerDeploymentResponse.status < 300) {
        const response = (await registerDeploymentResponse.json()) as RegisterDeploymentResponse;

        if (response?.services?.[0]?.name !== props.servicePath) {
          failureReason =
            "Restate service registration failed: service name indicated by service response" +
            ` ("${response?.services?.[0]?.name})) does not match the expected value ("${props.servicePath}")!`;
          console.error(failureReason);
          break; // don't throw immediately - let retry loop decide whether to abort
        }

        console.log("Success!");
        return;
      }
    } catch (e) {
      console.error(`Service registration call failed: ${(e as Error)?.message} (attempt ${attempt})`);
      failureReason = `Restate service registration failed: ${(e as Error)?.message}`;
    }

    if (attempt >= MAX_REGISTRATION_ATTEMPTS) {
      console.error(`Service registration failed after ${attempt} attempts.`);
      break;
    }
    attempt += 1;
    const waitTimeMillis = randomInt(2_000) + 2 ** attempt * 1_000; // 3s -> 6s -> 10s
    console.log(`Retrying registration after ${waitTimeMillis} ms...`);
    await sleep(waitTimeMillis);
  }

  throw new Error(failureReason ?? "Restate service registration failed. Please see logs for details.");
};

async function createAuthHeader(props: RegistrationProperties): Promise<Record<string, string>> {
  if (!props.authTokenSecretArn) {
    return {};
  }

  console.log(`Using bearer authentication token from secret ${props.authTokenSecretArn}`);
  const ssm = new SecretsManagerClient();
  const response = await ssm.send(
    new GetSecretValueCommand({
      SecretId: props.authTokenSecretArn,
    }),
  );

  console.log(`Successfully retrieved secret "${response.Name}" version ${response.VersionId}`);
  return {
    Authorization: `Bearer ${response.SecretString}`,
  };
}

async function sleep(millis: number) {
  await new Promise((resolve) => setTimeout(resolve, millis));
}
