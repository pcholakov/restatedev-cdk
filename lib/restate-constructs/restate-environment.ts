import * as iam from "aws-cdk-lib/aws-iam";
import * as cdk from "aws-cdk-lib";
import * as ssm from "aws-cdk-lib/aws-secretsmanager";

/**
 * A Restate environment is a distinct deployment target. These could be an isolated environment in Restate Cloud, or
 * a self-hosted deployment.
 */
export interface RestateEnvironment {
  readonly invokerRole: iam.IRole;
  readonly adminUrl: string;
  readonly authToken?: ssm.ISecret;
  readonly registrationProviderToken: cdk.CfnOutput;
}