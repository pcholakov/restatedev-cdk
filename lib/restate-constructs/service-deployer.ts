/*
 * Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */

import { Construct } from "constructs";
import * as ssm from "aws-cdk-lib/aws-secretsmanager";
import * as lambda_node from "aws-cdk-lib/aws-lambda-nodejs";
import path from "node:path";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as cdk from "aws-cdk-lib";
import * as cr from "aws-cdk-lib/custom-resources";

const DEFAULT_TIMEOUT = cdk.Duration.seconds(180);

/**
 * This construct implements a custom CloudFormation resource provider that handles deploying Lambda-based service
 * handlers with a Restate environment. It is used internally by the Cloud and self-hosted Restate environment
 * constructs and not intended for direct use by end users of Restate.
 *
 * This functionality is implemented as a custom resource so that we are notified of any updates to service handler
 * functions: by creating a CloudFormation component, we can model the dependency that any changes to the handlers need
 * to be communicated to the registrar. Without this dependency, CloudFormation might perform an update deployment that
 * triggered by a Lambda handler code or configuration change, and the Restate environment would be unaware of it.
 *
 * You can share the same instance across multiple service registries provided the configuration options are compatible
 * (e.g. the Restate environments it needs to communicate with for deployment are all accessible via the same VPC and
 * Security Groups).
 */
export class ServiceDeployer extends Construct {
  /** The custom resource provider for handling "deployment" resources. */
  readonly deploymentResourceProvider: cr.Provider;

  constructor(
    scope: Construct,
    id: string,
    props: {
      authToken?: ssm.ISecret;
    } & Pick<lambda.FunctionOptions, "functionName" | "logGroup" | "timeout" | "vpc" | "vpcSubnets" | "securityGroups">,
  ) {
    super(scope, id);

    const eventHandler = new lambda_node.NodejsFunction(this, "EventHandler", {
      functionName: props.functionName,
      logGroup: props.logGroup,
      description: "Restate custom registration handler",
      entry: path.join(__dirname, "register-service-handler/index.js"),
      architecture: lambda.Architecture.ARM_64,
      runtime: lambda.Runtime.NODEJS_LATEST,
      memorySize: 128,
      timeout: props.timeout ?? DEFAULT_TIMEOUT,
      environment: {
        NODE_OPTIONS: "--enable-source-maps",
      },
      bundling: {
        minify: false,
        sourceMap: true,
      },
      ...(props.vpc
        ? ({
            vpc: props.vpc,
            vpcSubnets: props.vpcSubnets,
            securityGroups: props.securityGroups,
          } satisfies Pick<lambda.FunctionOptions, "vpc" | "vpcSubnets" | "securityGroups">)
        : {}),
    });
    props.authToken?.grantRead(eventHandler);

    this.deploymentResourceProvider = new cr.Provider(this, "CustomResourceProvider", { onEventHandler: eventHandler });
  }
}
