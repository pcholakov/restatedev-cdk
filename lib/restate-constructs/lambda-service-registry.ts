/*
 * Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */

import * as cdk from "aws-cdk-lib";
import * as iam from "aws-cdk-lib/aws-iam";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { Construct } from "constructs";
import { RegistrationProperties } from "./register-service-handler";

import { RestateEnvironment } from "./restate-environment";

/**
 * A Restate RPC service path. Example: `greeter`.
 */
export type RestatePath = string;

/**
 * Manage registration of a set of Lambda-deployed Restate RPC Service handlers with a Restate environment.
 */
export type LambdaServiceRegistryProps = {
  /**
   * Custom resource provider token required for service discovery.
   */
  environment: RestateEnvironment;

  /**
   * Lambda Restate service handlers to deploy.
   */
  handlers: Record<RestatePath, lambda.Function>;
}

/**
 * Manage registration of a set of Lambda-deployed Restate RPC Service handlers with a Restate environment. This
 * construct creates a custom resource which will trigger Restate service discovery on handler function changes.
 */
export class LambdaServiceRegistry extends Construct {
  private readonly registrationProviderToken: cdk.CfnOutput;
  private readonly serviceHandlers: Record<RestatePath, lambda.Function>;

  constructor(scope: Construct, id: string, props: LambdaServiceRegistryProps) {
    super(scope, id);

    if (Object.values(props.handlers).length == 0) {
      throw new Error("Please specify at least one service handler.");
    }

    this.serviceHandlers = props.handlers;
    this.registrationProviderToken = props.environment.registrationProviderToken;
    this.registerServices(props.environment);
  }

  private registerServices(environment: RestateEnvironment) {
    const invokerRole = iam.Role.fromRoleArn(this, "InvokerRole", environment.invokerRole.roleArn);

    const allowInvokeFunction = new iam.Policy(this, "AllowInvokeFunction", {
      statements: [
        new iam.PolicyStatement({
          sid: "AllowInvokeAnyFunctionVersion",
          actions: ["lambda:InvokeFunction"],
          resources: Object.values(this.serviceHandlers)
            .map(handler => handler.functionArn + ":*"),
        }),
      ],
    });

    invokerRole.attachInlinePolicy(allowInvokeFunction);

    for (const [path, handler] of Object.entries(this.serviceHandlers)) {
      this.registerHandler({
        adminUrl: environment.adminUrl,
        invokerRoleArn: invokerRole.roleArn,
        authTokenSecretArn: environment.authToken?.secretArn,
      }, { path, handler }, allowInvokeFunction);
    }
  }

  private registerHandler(restate: EnvironmentDetails, service: {
    path: RestatePath,
    handler: lambda.Function
  }, allowInvokeFunction: iam.Policy) {
    const registrar = new RestateServiceRegistrar(this, service.handler.node.id + "Discovery", {
      environment: restate,
      service,
      serviceToken: this.registrationProviderToken.value,
    });

    // CloudFormation doesn't know that Restate depends on this role to call services; we must ensure that Lambda
    // permission changes are applied before we can trigger discovery (represented by the registrar).
    registrar.node.addDependency(allowInvokeFunction);
  }
}

class RestateServiceRegistrar extends Construct {
  constructor(scope: Construct, id: string,
              props: {
                environment: EnvironmentDetails,
                service: {
                  path: RestatePath,
                  handler: lambda.Function
                },
                serviceToken: string,
              },
  ) {
    super(scope, id);

    new cdk.CustomResource(this, props.service.handler.node.id + "Discovery", {
      serviceToken: props.serviceToken,
      resourceType: "Custom::RestateServiceRegistrar",
      properties: {
        servicePath: props.service.path,
        adminUrl: props.environment.adminUrl,
        authTokenSecretArn: props.environment.authTokenSecretArn,
        serviceLambdaArn: props.service.handler.currentVersion.functionArn,
        invokeRoleArn: props.environment.invokerRoleArn,
        removalPolicy: cdk.RemovalPolicy.RETAIN,
      } satisfies RegistrationProperties,
    });
  }
}

interface EnvironmentDetails {
  readonly adminUrl: string;
  readonly invokerRoleArn: string;
  readonly authTokenSecretArn?: string;
}