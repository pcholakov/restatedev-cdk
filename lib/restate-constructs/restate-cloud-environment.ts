/*
 * Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */

import { Construct } from "constructs";
import * as cdk from "aws-cdk-lib";
import * as iam from "aws-cdk-lib/aws-iam";
import * as ssm from "aws-cdk-lib/aws-secretsmanager";
import { RestateEnvironment } from "./restate-environment";
import { RegistrationProvider } from "./registration-provider";

const RESTATE_INGRESS_PORT = 8080;
const RESTATE_ADMIN_PORT = 9070;

export interface RestateCloudEnvironmentProps {
  /** Prefix for resources created by this construct that require unique names. */
  prefix?: string;

  /** ID of the Restate service cluster to which this service will be registered. */
  clusterId: string;

  /** Auth token for Restate environment. Used with the admin service for service deployment registration. */
  authTokenSecretArn: string;
}

/**
 * Restate environment hosted on Restate Cloud. This construct manages the role in the deployment environment that
 * Restate Cloud assumes to call registered services, and provides the service registration helper for Lambda-based
 * handlers. An appropriate trust policy will be added to this role that allows Restate to assume it from outside the
 * deployment AWS account.
 */
export class RestateCloudEnvironment extends Construct implements RestateEnvironment {
  readonly invokerRole: iam.Role;
  readonly ingressUrl: string;
  readonly adminUrl: string;
  readonly authToken: ssm.ISecret;
  readonly registrationProviderToken: cdk.CfnOutput;

  constructor(scope: Construct, id: string, props: RestateCloudEnvironmentProps) {
    super(scope, id);

    // This role should be easier to customize or override completely: https://github.com/restatedev/cdk/issues/21
    this.invokerRole = new iam.Role(this, "RestateServiceInvokerRole", {
      description: "Role assumed by Restate Cloud when invoking Lambda service handlers",
      assumedBy: new iam.ArnPrincipal("arn:aws:iam::663487780041:role/restate-dev"),
      externalIds: [props.clusterId],
    });

    this.ingressUrl = `https://${props.clusterId}.dev.restate.cloud:${RESTATE_INGRESS_PORT}`;
    this.adminUrl = `https://${props.clusterId}.dev.restate.cloud:${RESTATE_ADMIN_PORT}`;
    this.authToken = ssm.Secret.fromSecretCompleteArn(this, "ClusterAuthToken", props.authTokenSecretArn);

    const registrationProvider = new RegistrationProvider(this, "RegistrationProvider", { authToken: this.authToken });
    this.registrationProviderToken = new cdk.CfnOutput(this, "RegistrationProviderToken", {
      description: "Restate service registration provider custom component token used by registry to perform discovery",
      exportName: [props.prefix, "RegistrationProviderToken"].filter(Boolean).join("-"),
      value: registrationProvider.serviceToken,
    });
  }
}