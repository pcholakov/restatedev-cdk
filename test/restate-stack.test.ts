import * as cdk from "aws-cdk-lib";
import { RestateEnvironment, ServiceDeployer } from "../lib/restate-constructs";
import { Construct } from "constructs";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as secretsmanager from "aws-cdk-lib/aws-secretsmanager";
import * as iam from "aws-cdk-lib/aws-iam";
import "jest-cdk-snapshot";

describe("Restate constructs", () => {
  test("Deploy a Lambda service handler to a remote Restate environment", () => {
    const app = new cdk.App();
    const stack = new LambdaServiceDeployment(app, "LambdaServiceDeployment", {});

    expect(stack).toMatchCdkSnapshot({
      ignoreAssets: true,
      yaml: true,
    });
  });
});

class LambdaServiceDeployment extends cdk.Stack {
  constructor(scope: Construct, id: string, props: {} & cdk.StackProps) {
    super(scope, id, props);

    const invokerRole = new iam.Role(this, "InvokerRole", { assumedBy: new iam.AccountRootPrincipal() });

    const authToken = new secretsmanager.Secret(this, "RestateApiKey", {
      secretStringValue: cdk.SecretValue.unsafePlainText("api-key"),
    });

    const restateEnvironment = RestateEnvironment.fromAttributes({
      invokerRole,
      adminUrl: "https://restate.example.com:9070",
      authToken,
    });

    const handler: lambda.Function = new lambda.Function(this, "RestateServiceHandler", {
      runtime: lambda.Runtime.NODEJS_LATEST,
      handler: "index.handler",
      code: lambda.Code.fromInline("{ ... }"),
    });
    handler.grantInvoke(invokerRole);

    const serviceDeployer = new ServiceDeployer(this, "ServiceDeployer", {
      // only needed in testing, where the relative path of the registration function is different from how customers would use it
      entry: "dist/register-service-handler/index.js",
    });
    serviceDeployer.deployService("Service", handler.currentVersion, restateEnvironment);
  }
}
