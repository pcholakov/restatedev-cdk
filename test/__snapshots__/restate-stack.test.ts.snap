// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Restate constructs Deploy a Lambda service handler to a remote Restate environment 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "InvokerRole4DB2757E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "InvokerRoleDefaultPolicy713FD858": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "RestateServiceHandler71409CD7",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "RestateServiceHandler71409CD7",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "InvokerRoleDefaultPolicy713FD858",
        "Roles": [
          {
            "Ref": "InvokerRole4DB2757E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RestateApiKey6463672F": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "SecretString": "api-key",
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
    "RestateServiceHandler71409CD7": {
      "DependsOn": [
        "RestateServiceHandlerServiceRole07B26D05",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "{ ... }",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "RestateServiceHandlerServiceRole07B26D05",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "RestateServiceHandlerCurrentVersion40030E671fc2ba09c2d7b4ea8c6a3f8fee895a65": {
      "Properties": {
        "FunctionName": {
          "Ref": "RestateServiceHandler71409CD7",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "RestateServiceHandlerCurrentVersionRestateDeploymentE8F102EB": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "InvokerRoleDefaultPolicy713FD858",
        "InvokerRole4DB2757E",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "ServiceDeployerCustomResourceProviderframeworkonEvent528FE6C2",
            "Arn",
          ],
        },
        "adminUrl": "https://restate.example.com:9070",
        "authTokenSecretArn": {
          "Ref": "RestateApiKey6463672F",
        },
        "invokeRoleArn": {
          "Fn::GetAtt": [
            "InvokerRole4DB2757E",
            "Arn",
          ],
        },
        "private": "false",
        "removalPolicy": "retain",
        "serviceLambdaArn": {
          "Ref": "RestateServiceHandlerCurrentVersion40030E671fc2ba09c2d7b4ea8c6a3f8fee895a65",
        },
        "servicePath": "Service",
      },
      "Type": "Custom::RestateServiceDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "RestateServiceHandlerServiceRole07B26D05": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ServiceDeployerCustomResourceProviderframeworkonEvent528FE6C2": {
      "DependsOn": [
        "ServiceDeployerCustomResourceProviderframeworkonEventServiceRoleDefaultPolicy740A65C9",
        "ServiceDeployerCustomResourceProviderframeworkonEventServiceRole865AFB0C",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "7382a0addb9f34974a1ea6c6c9b063882af874828f366f5c93b2b7b64db15c94.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (LambdaServiceDeployment/ServiceDeployer/CustomResourceProvider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "ServiceDeployerEventHandler89EAD25F",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "ServiceDeployerCustomResourceProviderframeworkonEventServiceRole865AFB0C",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ServiceDeployerCustomResourceProviderframeworkonEventServiceRole865AFB0C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ServiceDeployerCustomResourceProviderframeworkonEventServiceRoleDefaultPolicy740A65C9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "ServiceDeployerEventHandler89EAD25F",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "ServiceDeployerEventHandler89EAD25F",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ServiceDeployerCustomResourceProviderframeworkonEventServiceRoleDefaultPolicy740A65C9",
        "Roles": [
          {
            "Ref": "ServiceDeployerCustomResourceProviderframeworkonEventServiceRole865AFB0C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ServiceDeployerEventHandler89EAD25F": {
      "DependsOn": [
        "ServiceDeployerEventHandlerServiceRoleDefaultPolicyFE2DC3C9",
        "ServiceDeployerEventHandlerServiceRoleF133584F",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e4ec531617dd4885872bb89a12c5f0447bf954ef5c73fc92841a978a48428ce5.zip",
        },
        "Description": "Restate custom registration handler",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "ServiceDeployerEventHandlerServiceRoleF133584F",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 180,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ServiceDeployerEventHandlerServiceRoleDefaultPolicyFE2DC3C9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": {
                "Ref": "RestateApiKey6463672F",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ServiceDeployerEventHandlerServiceRoleDefaultPolicyFE2DC3C9",
        "Roles": [
          {
            "Ref": "ServiceDeployerEventHandlerServiceRoleF133584F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ServiceDeployerEventHandlerServiceRoleF133584F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
